SARAALERT-832: Ability to delete and edit their own user comments

ashankland opened the pull request:
# Description
Jira Ticket: [SARAALERT-832](https://tracker.codev.mitre.org/browse/SARAALERT-832)

Provides a simple inline interface for users to edit or delete their own comments on patient records. 

Details of the implementation include: 
- For delete, don't actually delete the history item.  Add deleted_by and deleted_reason fields (that can be used to tell what is deleted)
- For edit, don't overwrite existing comment, but instead create a new comment that stores the original comment's id, providing a paper trail of all comment edits
- For the UI display, only display the most recent version of non-deleted history comments.  This is done by filtering only histories that have `deleted_by` as nil.  If there are duplicate `original_comment_id` (meaning there are multiple versions), only the most recent is shown
- For export, export ALL histories, allowing the user to construct a paper trail for audit if needed

## (Feature) Demo/Screenshots
![Screen Shot 2021-04-22 at 9 38 59 AM](https://user-images.githubusercontent.com/35042815/115724055-98b97680-a34e-11eb-8a22-295173d6826d.png)

![Screen Shot 2021-04-15 at 3 36 27 PM](https://user-images.githubusercontent.com/35042815/114928372-872d1780-9e00-11eb-8455-bf28f560f9a3.png)

## Important Changes

Added three new db fields to histories (original_comment_id, deleted_by and deleted_reason): `schema.rb`, `history.rb`, `20210430124303_add_edited_and_deleted_to_histories.rb`

Added support for "deleting" a history: `History.js`, `histories_controller.rb`, `user.rb`

Added support for "editing" a history: `History.js`, `histories_controller.rb`, `user.rb`

Updated the UI to only show non deleted and the latest version of histories: `HistoryList.js`, `patients_controller.rb`, `show.html.erb`

Added edited flag to denote if a history comment has been edited: `History.js`, `bootstrap.scss`

Updated export fields to include the 3 new ones plus the history id for edit comparison: `import_export_constants.rb`

Added edit and delete routes for histories: `histories_controller.rb`, `routes.rb`

Moved history components into their own folder in `/patient`: `History.js`, `HistoryList.js`

Renamed `HistoryComponent` to match our other table components : `HistoryList.js`, `show.html.erb`

Added frontend tests for history components: `History.test.js`, `HistoryList.test.js`, `mockHistories.js`, `mockUsers.js`

Added new generic delete dialog component and frontend tests: `History.js`, `DeleteDialog.js`, `DeleteDialog.test.js`

Moved history relative date formatter to the DateTime util file: `History.js`, `Datetime.js`, `helpers.js`

Removed unused history erb file: `_histories.html.erb`

Added for a previous version of this ticket but I think it's worth keeping for future use, added support for custom variant on submit button of confirm dialog: `ConfirmDialog.js`

# Checklists

**Submitter:**
- [x] This PR describes why these changes were made.
- [x] This PR is into the correct branch.
- [x] This PR includes the correct JIRA ticket reference.
- [x] Code diff has been reviewed (it **does not** contain: additional white space, not applicable code changes, debug statements, etc.)
- [x] If UI changes have been made, Chrome Dev Tools Lighthouse accessibility test has been executed to ensure no 508 issues were introduced.
- [x] Tests are included and test edge cases
- [x] Tests have been run locally and pass (remember to update Gemfile when applicable)
- [ ] Test fixtures updated and documented as necessary


@holmesie  :
- [x] Code is maintainable and reusable, reuses existing code and infrastructure where appropriate, and accomplishes the task’s purpose
- [x] The tests appropriately test the new code, including edge cases
- [x] You have tried to break the code
- [x] If applicable, you have tested changes against a large database, and considered possible performance regressions


@tstrass  :
- [x] Code is maintainable and reusable, reuses existing code and infrastructure where appropriate, and accomplishes the task’s purpose
- [x] The tests appropriately test the new code, including edge cases
- [x] You have tried to break the code
- [x] If applicable, you have tested changes against a large database, and considered possible performance regressions


@DPC15038  :
- [x] Code is maintainable and reusable, reuses existing code and infrastructure where appropriate, and accomplishes the task’s purpose
- [x] The tests appropriately test the new code, including edge cases
- [x] You have tried to break the code
- [x] If applicable, you have tested changes against a large database, and considered possible performance regressions

sgober commented on app/javascript/components/history/History.js:
```diff
@@ -1,31 +1,134 @@
+    };
+  }
+
+  handleEditComment = () => {
```
I would call this maybe 'showEdit' or something

sgober commented on app/javascript/components/history/History.js:
```diff
@@ -1,31 +1,134 @@
+    this.setState({ comment: event.target.value });
+  };
+
+  handleDeleteComment = () => {
```
I just noticed that there is no confirmation for delete... I think there needs to be a confirmation when you click the delete button

sgober replied:
You can use the confirmDialog component, would be super simple

sgober reviewed (changes requested):
The styling behavior isn't quite right (see below in the screenshot for reference)
- There is some weird padding going on next to the badge on comments that have the two new buttons.  This badge should be aligned with the badges of the other history messages
- I don't love how the placement of the edit/delete buttons vertically off centers the user name in the badge - could we move them to the left of the badge?  not sure the best solution here but I would consult with Mike Nosal
- When you click the edit button, the previous edit text is a placeholder and not actually populated in the dropdown.  This is a problem because users want to edit the existing text, and the way this currently works they have to start from scratch
- There should be a cancel button for the edit option
- I said this in my inline comments, but the delete option should have a confirmation dialog

![Screen Shot 2021-04-02 at 8 04 57 AM](https://user-images.githubusercontent.com/35042815/113414260-9afd6600-938a-11eb-96f1-e55b35936bc8.png)

Some additional comments...
1. Quite a few system tests are failing, this needs to be addressed
2. There is also a brakeman error in the Static Analysis
3. Since the History component was revamped, I'd like to see front end tests written for it

sgober commented on app/javascript/components/history/History.js:
```diff
@@ -1,31 +1,134 @@
+    });
+  };
+
+  renderEditMode = () => {
```
This is confusing because this renders both edit mode and not, I would remove the conditional and only have this render the edit stuff.  Then add a CardBody in the render below with a conditional that renders either `Card.Text>{this.props.history.comment}</Card.Text>` or `renderEdit`

you also can use `this.state.edit_mode` in the conditional

sgober commented on app/javascript/components/history/History.js:
```diff
@@ -1,31 +1,134 @@
+              <div className="badge-padding h5">
+                <span className="badge badge-secondary">{this.props.history.history_type}</span>
+              </div>
+              {this.formatEditable()}
```
Add `{this.props.history.history_type === 'Comment' && this.formatEditable()` here and remove the conditional from the `formatEditable` method

sgober commented:
> The styling behavior isn't quite right (see below in the screenshot for reference)
> 
> * I don't love how the placement of the edit/delete buttons vertically off centers the user name in the badge - could we move them to the left of the badge?  not sure the best solution here but I would consult with Mike Nosal. Maybe these turn into buttons in the card body?

![Screen Shot 2021-04-02 at 8 41 07 AM](https://user-images.githubusercontent.com/35042815/113416414-490b0f00-938f-11eb-9c08-227532105d92.png)

Something like this would be good, obviously the buttons would have to be cleaned up and what not.  It would be nice then too if the edit comment button showed up here as well when in edit mode, hiding the other buttons

sgober commented on app/javascript/components/history/History.js:
```diff
@@ -1,31 +1,134 @@
+    });
+  };
+
+  formatEditable = () => {
```
since this includes the delete button as well, I'd call this `renderHistoryActionButtons` or something similar

sgober commented on app/javascript/components/history/History.js:
```diff
@@ -1,31 +1,134 @@
+  constructor(props) {
+    super(props);
+    this.state = {
+      edit_mode: false,
```
one small thing... can you make this camel case?

sgober commented on app/javascript/components/history/History.js:
```diff
@@ -1,31 +1,134 @@
+          <Card.Header>
+            <b>{this.props.history.created_by}</b>, {time_ago_in_words(moment(this.props.history.created_at).toDate())} ago (
+            {moment
+              .tz(this.props.history.created_at, 'UTC')
```
An additional thought, can we move this time_ago_in_words method (plus the ago text) into the /javascript/utils/DateTime.js and call it `formatRelativeDate`? @DPC15038 started breaking out our util methods into one place and I think this would be a good candidate

ashankland commented:
@sgober this should be good for a re-review. I've moved the edit and delete buttons to a footer, made them buttons rather than links, added a cancel option and a delete confirmation popup, plus some general renaming. I'm honestly not completely sold on the footer buttons because it looks weird to me to have a large footer only on some histories and not others, but I defer to you there. 

I've also looked into the static analysis issue and it looks like Brakeman is confused: it thinks that because the History.find_by_id function isn't scoped, you could get arbitrary comments, but the redirect above prevents anyone who can't create or edit histories from accessing any, and the redirect below prevents the functions from returning any data at all unless the caller is logged in as the creator of that history item.

sgober commented on app/javascript/tests/history/History.test.js:
```diff
@@ -0,0 +1,57 @@
+
+const authyToken = 'Q1z4yZXLdN+tZod6dBSIlMbZ3yWAUFdY44U06QWffEP76nx1WGMHIz8rYxEUZsl9sspS3ePF2ZNmSue8wFpJGg==';
+
+beforeEach(() => {
```
Is there a reason this is in here? I think it's unnecessary

sgober commented on app/javascript/tests/history/HistoryList.test.js:
```diff
@@ -0,0 +1,26 @@
+const goToMock = jest.fn();
+const authyToken = 'Q1z4yZXLdN+tZod6dBSIlMbZ3yWAUFdY44U06QWffEP76nx1WGMHIz8rYxEUZsl9sspS3ePF2ZNmSue8wFpJGg==';
+
+beforeEach(() => {
```
Same here

sgober commented on app/javascript/tests/history/HistoryList.test.js:
```diff
@@ -0,0 +1,26 @@
+import { mockHistory1, mockHistory2 } from '../mocks/mockHistories'
+import { nameFormatter, formatDate } from '../util.js'
+
+const goToMock = jest.fn();
```
this variable is unused and needs to be removed

sgober commented on app/javascript/tests/mocks/mockHistories.js:
```diff
@@ -0,0 +1,25 @@
+
+const mockHistory1 = {
+  id: 1,
```
I would add `was_edited: true` to one of these and test the edited tag functionality

sgober commented on app/javascript/tests/history/History.test.js:
```diff
@@ -0,0 +1,57 @@
+describe('History', () => {
+  it('Properly renders non-comment histories', () => {
+    const wrapper = shallow(<History key={mockHistory1.id} history={mockHistory1} authenticity_token={authyToken} />);
+    const divEditId = "edit-history-item-" + mockHistory1.id;
```
can you append the `#` here so you don't have to do it on lines 22 and 23?

sgober commented on app/javascript/tests/history/History.test.js:
```diff
@@ -0,0 +1,57 @@
+
+  it('Properly renders comment histories', () => {
+    const wrapper = shallow(<History key={mockHistory2.id} history={mockHistory2} authenticity_token={authyToken} />);
+    const divEditId = "edit-history-item-" + mockHistory2.id;
```
All the same comments about the prior test apply to this one as well

sgober replied:
In addition, I would also check that `find(`ReactTooltip`).length === 2` here and make sure the tooltip text is accurate

sgober commented on app/javascript/tests/history/History.test.js:
```diff
@@ -0,0 +1,57 @@
+    expect(wrapper.find('#' + divDeleteId).exists()).toBeTruthy();
+    expect(wrapper.find('[data-for="'+divDeleteId+'"]').exists()).toBeTruthy();
+  });
+});
```
I would also like to see tests for
- Changing the text in the edit field properly updates the state
- The disabling functionality of the update/cancel buttons when in edit mode (i.e. when text is empty)
- Clicking the update button calls `handleEditSubmit`
- Clicking the delete button calls `delete` (unfortunately can't test much more than this because of the confirmation dialog)
- Properly renders the `(edited)` tag when `was_edited` is true

sgober replied:
added here: https://github.com/SaraAlert/SaraAlert/pull/793/commits/336f7f2565ad7b9798767f02dd44c3442b64135d

sgober commented on app/javascript/tests/history/History.test.js:
```diff
@@ -0,0 +1,57 @@
+    const divEditId = "edit-history-item-" + mockHistory1.id;
+    const divDeleteId = "delete-history-item-" + mockHistory1.id;
+    expect(wrapper.find('b').exists()).toBeTruthy();
+    expect(wrapper.find('b').text()).toEqual(mockHistory1.created_by);
```
Please add checks for the relative date and timestamp, as well as the history text contained in the card body

sgober commented on app/javascript/tests/history/History.test.js:
```diff
@@ -0,0 +1,57 @@
+    expect(wrapper.find('b').exists()).toBeTruthy();
+    expect(wrapper.find('b').text()).toEqual(mockHistory1.created_by);
+    expect(wrapper.find('.h5').find('span').text()).toEqual(mockHistory1.history_type);
+    expect(wrapper.find('#' + divEditId).exists()).toBeFalsy();
```
so this is actually testing to see if the tooltip is not present, when we want to test that the buttons . I would do a find Button and make sure the length is 0.  For the following test make sure the length is 2 and then test each button and make sure the icon is correct (trash and edit)

sgober replied:
fixed here: https://github.com/SaraAlert/SaraAlert/pull/793/commits/336f7f2565ad7b9798767f02dd44c3442b64135d

sgober commented on app/javascript/tests/history/History.test.js:
```diff
@@ -0,0 +1,57 @@
+    const divDeleteId = "delete-history-item-" + mockHistory2.id;
+    expect(wrapper.find('b').text()).toEqual(mockHistory2.created_by);
+    expect(wrapper.find('#' + divEditId).exists()).toBeTruthy();
+    expect(wrapper.find('[data-for="'+divEditId+'"]').exists()).toBeTruthy();
```
I would find these elements like I said above.  Also be sure to check the button and tooltip text

sgober commented on app/javascript/tests/history/History.test.js:
```diff
@@ -0,0 +1,57 @@
+  it('Clicking edit properly displays the field', () => {
+    const wrapper = shallow(<History key={mockHistory2.id} history={mockHistory2} authenticity_token={authyToken} />);
+    const divEditId = "edit-history-item-" + mockHistory2.id;
+    wrapper.find('[data-for="'+divEditId+'"]').find(Button).simulate('click');
```
same here about the find for the click

sgober commented on app/javascript/tests/history/History.test.js:
```diff
@@ -0,0 +1,57 @@
+    const wrapper = shallow(<History key={mockHistory2.id} history={mockHistory2} authenticity_token={authyToken} />);
+    const divEditId = "edit-history-item-" + mockHistory2.id;
+    wrapper.find('[data-for="'+divEditId+'"]').find(Button).simulate('click');
+    expect(wrapper.find('#comment').exists()).toBeTruthy();
```
here you also want to test for the update/cancel buttons (same way as I suggested above) as well as that the text in the input is correct

sgober replied:
added here: https://github.com/SaraAlert/SaraAlert/pull/793/commits/336f7f2565ad7b9798767f02dd44c3442b64135d

sgober commented on app/javascript/tests/history/History.test.js:
```diff
@@ -0,0 +1,57 @@
+    expect(wrapper.find('[data-for="'+divDeleteId+'"]').exists()).toBeTruthy();
+  });
+
+  it('Clicking edit properly displays the field', () => {
```
I would change this to `Clicking the edit button properly updates state and displays edit mode` and also add a check that editMode is false at the beginning and true at the end

sgober commented on app/javascript/tests/history/History.test.js:
```diff
@@ -0,0 +1,57 @@
+    expect(wrapper.find('#comment').exists()).toBeTruthy();
+  });
+
+  it('Canceling edit properly restores original view', () => {
```
`Clicking the cancel button reverts from edit mode to original view`

sgober replied:
All of my above comments apply to this test as well

sgober commented on app/javascript/tests/history/History.test.js:
```diff
@@ -0,0 +1,57 @@
+    const divEditId = "edit-history-item-" + mockHistory2.id;
+    const divDeleteId = "delete-history-item-" + mockHistory2.id;
+    wrapper.find('[data-for="'+divEditId+'"]').find(Button).simulate('click');
+    wrapper.find('[aria-label="Cancel Edit History Comment"]').simulate('click');
```
in between these two clicks check that you are indeed in edit mode (state and check html)

sgober replied:
added here: https://github.com/SaraAlert/SaraAlert/pull/793/commits/336f7f2565ad7b9798767f02dd44c3442b64135d

sgober commented on app/javascript/tests/history/HistoryList.test.js:
```diff
@@ -0,0 +1,26 @@
+describe('HistoryList', () => {
+  it('Properly renders all history components', () => {
+    const wrapper = shallow(<HistoryList patient_id={mockPatient1.id} histories={[]} authenticity_token={authyToken} 
+      history_types={{enrollment: "Enrollment", comment: "Comment"}}/>);
```
This test is missing a lot of key elements on the history page.  I would pass in `[mockHistory1, mockHistory2` and test that there are two history components. In addition, be sure to test that every dropdown, button, input, tooltips, child components etc are all there

sgober replied:
added here: https://github.com/SaraAlert/SaraAlert/pull/793/commits/336f7f2565ad7b9798767f02dd44c3442b64135d

sgober commented on app/javascript/tests/history/HistoryList.test.js:
```diff
@@ -0,0 +1,26 @@
+    expect(wrapper.find('#histories').find('span').text()).toEqual('History ');
+    expect(wrapper.find('#histories').find('textarea').exists()).toBeTruthy();
+  });
+});
```
This component has additional functionality that needs to be tested.  I would like to see tests for
- The two selects call the right functions and properly update the state
- The list is properly filtered
- Pagination
- Changing text in the input properly updates state
- The add comment button is disabled when it should be
- Clicking the add comment button calls the right function

sgober replied:
added here: https://github.com/SaraAlert/SaraAlert/pull/793/commits/336f7f2565ad7b9798767f02dd44c3442b64135d

sgober reviewed (changes requested):
This is a good start, but there are quite a few test cases that are missing and cleanup that needs to be done

timwongj commented on app/lib/import_export_constants.rb:
```diff
@@ -373,7 +373,8 @@ module ImportExportConstants # rubocop:todo Metrics/ModuleLength
     created_at: 'History Created Date',
-    updated_at: 'History Updated Date'
+    updated_at: 'History Updated Date',
+    was_edited: 'Has Been Edited'
```
is this language ('Has Been Edited') confirmed with the PO team?

sgober replied:
@ashankland?

ashankland replied:
It has not: the was_edited field got added after this item was ran by the PO team I believe.

sgober replied:
also obsolete

timwongj commented on app/javascript/packs/stylesheets/bootstrap.scss:
```diff
@@ -31,6 +31,15 @@ body {
   border-radius: 0;
 }
 
+.icon-btn {
```
are these specific to the history component? if so, maybe we should put it in a separate stylesheet for histories and give it a separate class

sgober replied:
right now they're just in the history component, however I made it generic in case we ever want to add these types of buttons elsewhere

sgober commented:
@holmesie @tstrass @DPC15038 this should be good for review

timwongj commented on app/lib/import_export_constants.rb:
```diff
@@ -522,7 +523,7 @@ def self.rct_node(schema, label, fields)
     label: 'Histories',
     nodes: [rct_node(:histories, 'History', %i[patient_id user_defined_id_statelocal user_defined_id_cdc user_defined_id_nndss id created_by history_type
-                                               comment created_at updated_at])]
+                                               comment created_at updated_at was_edited])]
```
let's move `was_edited` to be before the timestamps, same for line 378

sgober replied:
done here: https://github.com/SaraAlert/SaraAlert/pull/793/commits/ad9df046c5344445c1cf056f8dfb316aee18461a

sgober replied:
also now obsolete

timwongj commented on app/javascript/components/history/helpers.js:
```diff
@@ -1,41 +0,0 @@
-export function time_ago_in_words(date) {
```
looks like `time_ago_in_words` is still being referenced in `_histories.html.erb`, what's the reasoning for unrefactoring this function from a separate file?

sgober replied:
We started putting all of our DateTime util functions in a separate file, so it was moved there.  I can update the reference

sgober replied:
looks like this file isn't even being used anymore, so I went ahead and removed it here: https://github.com/SaraAlert/SaraAlert/pull/793/commits/ad9df046c5344445c1cf056f8dfb316aee18461a

tstrass commented on app/lib/import_export_constants.rb:
```diff
@@ -522,7 +523,7 @@ def self.rct_node(schema, label, fields)
     label: 'Histories',
     nodes: [rct_node(:histories, 'History', %i[patient_id user_defined_id_statelocal user_defined_id_cdc user_defined_id_nndss id created_by history_type
-                                               comment created_at updated_at])]
+                                               comment was_edited created_at updated_at])]
```
This adds 'Edited' in custom export, but not full exports. Please also add this to the `FULL_HISTORY_HISTORIES_FIELDS` and `FULL_HISTORY_HISTORIES_HEADERS` (unless there was a deliberate decision I missed to not do that).

timwongj replied:
Kelly just confirmed to add this to full history exports as well as the last field after created/updated at

sgober replied:
this has been removed for now as the fields will be changing, but will add back with the right fields once implemented

sgober replied:
This has been updated

tstrass commented on db/schema.rb:
```diff
@@ -401,6 +402,8 @@
     t.boolean "race_refused_to_answer"
     t.boolean "latest_assessment_symptomatic", default: false
     t.date "first_positive_lab_at"
+    t.string "legacy_primary_language"
+    t.string "legacy_secondary_language"
```
This looks like an artifact of the 1344 migration. You need to roll that back locally when switch branches.

sgober replied:
resolved

DPC15038 commented on app/javascript/components/patient/history/History.js:
```diff
@@ -0,0 +1,193 @@
+            <span>You may edit comments you have added</span>
+          </ReactTooltip>
+          <span data-for={`delete-history-item-${this.props.history.id}`} data-tip="">
+            <Button id="delete-history-btn" variant="link" className="icon-btn p-0" onClick={this.handleDeleteClick} aria-label="Delete History Comment">
```
if `handleDeleteClick` is just calling `this.delete` then why not change the `onClick` action to `() => this.delete()`.

sgober replied:
This comment is obsolete

DPC15038 commented on app/javascript/components/patient/history/HistoryList.js:
```diff
@@ -1,16 +1,16 @@
 import _ from 'lodash';
 import { PropTypes } from 'prop-types';
-import { Card, Row } from 'react-bootstrap';
+import { Button, Card, Row } from 'react-bootstrap';
```
Really small unimportant thing, but I appreciate the alphabetizing here. It keeps our code base neat and tidy.

sgober commented:
Hold off on reviewing this for now, the functionality is going to get an overhaul

sgober commented on app/lib/import_export_constants.rb:
```diff
@@ -374,7 +374,8 @@ module ImportExportConstants # rubocop:todo Metrics/ModuleLength
     created_at: 'History Created Date',
-    updated_at: 'History Updated Date'
+    updated_at: 'History Updated Date',
+    was_edited: 'Has Been Edited'
```
this is now obsolete... resolving

sgober commented:
@holmesie @tstrass @DPC15038 @timwongj - this is all set to be reviewed

@mayerm94, @ashankland tagging you for your reference

tstrass reviewed (changes requested):
Seeing this
![image](https://user-images.githubusercontent.com/8235974/116939919-20e72800-ac3b-11eb-8c71-9244b0544191.png)

DPC15038 commented on app/controllers/histories_controller.rb:
```diff
@@ -8,9 +8,50 @@ class HistoriesController < ApplicationController
+    # mark each version of the history as deleted, not just the most recent one
+    history_versions = patient.histories.where(original_comment_id: history.original_comment_id)
+
+    history_versions.each do |h|
```
Rather than iterating over each `history_version` here, why not use [update_all](https://apidock.com/rails/ActiveRecord/Relation/update_all)? I haven't tested this, but maybe something like
```suggestion
    patient.histories.where(original_comment_id: history.original_comment_id).update_all({deleted_by: current_user.email, delete_reason: params.permit(:delete_reason)[:delete_reason]})
```

sgober replied:
good idea! changed here: https://github.com/SaraAlert/SaraAlert/pull/793/commits/fb980cdbbffbb305ed63c0cb7a76874d8b185116

DPC15038 commented on app/javascript/components/util/DeleteDialog.js:
```diff
@@ -0,0 +1,85 @@
+        </Modal.Header>
+        <Modal.Body>
+          <p>
+            Are you sure you want to delete this {this.props.type}?&nbsp; This action cannot be undone.&nbsp; For auditing purposes, this deletion will be
```
Why the `&nbsp;` here? if you need, you can insert an empty space `{' '}` for line break issues? This currently is causing two spaces

https://user-images.githubusercontent.com/17532163/117021810-add6c380-acc5-11eb-82cf-ded2be1d183e.mov

sgober replied:
fixed here: https://github.com/SaraAlert/SaraAlert/pull/793/commits/fb980cdbbffbb305ed63c0cb7a76874d8b185116

DPC15038 commented on app/javascript/components/util/DeleteDialog.js:
```diff
@@ -0,0 +1,85 @@
+          <Button variant="danger btn-square" onClick={this.delete} disabled={this.state.disabled || this.state.loading}>
+            {this.state.loading && (
+              <React.Fragment>
+                <span className="spinner-border spinner-border-sm" role="status" aria-hidden="true"></span>&nbsp;
```
same question here with the `&nbsp;`? You could just put a white-space between the `<span>`s

DPC15038 commented on app/javascript/components/patient/history/History.js:
```diff
@@ -0,0 +1,185 @@
+          variant="primary"
+          size="sm"
+          className="float-right mt-2"
+          disabled={this.state.loading || this.state.comment === '' || this.state.comment === this.props.history.comment}
```
I really like this bit of the conditional `this.state.comment === this.props.history.comment`. 👍 👍

DPC15038 commented on app/javascript/components/patient/history/History.js:
```diff
@@ -0,0 +1,185 @@
+  renderEditMode() {
+    return (
+      <React.Fragment>
+        <textarea
```
Do we want to add a limit to this? I believe SQL only allows strings of 8,000 characters. The call fails if you try to send more than that to the backend.

sgober replied:
that would be a question for @holmesie.  currently we don't cap in the add comment text input so if we do one we should do the other

holmesie replied:
We did at some point, but it looks like that was lost when this was converted to a React component.  This field is actually not a `string` in the db, it's a `text` - so it does support more than 8k. We limit it to 10k characters in the model validation. This is great feedback, and it's never a bad idea to add that extra bit of validation to the front end.

sgober replied:
So what do you want to limit this at?

holmesie replied:
Well, we limit it to 10k characters on the backend, so 10k is probably a good choice here as well.

sgober replied:
Added here: https://github.com/SaraAlert/SaraAlert/pull/793/commits/ddeeb2c07f876d766273b0a1823b00177975d4e4

holmesie reviewed (changes requested):
This needs to allow users to see older versions of comments (it's okay to rely on exports for archived, but not for edited comments). I would look to GitHub comment edits for inspiration as to how it should be done.

timwongj commented on config/routes.rb:
```diff
@@ -65,7 +65,10 @@
   post 'admin/reset_2fa', to: 'admin#reset_2fa'
   post 'admin/email_all', to: 'admin#email_all'
 
-  resources :histories, only: [:create]
```
Can we setup these routes the rails way like [this](https://guides.rubyonrails.org/routing.html#crud-verbs-and-actions) instead?
```
resources :histories, only: [:create, :update, :destroy]
```

timwongj replied:
@holmesie I spoke with @sgober about this and she mentioned that it wouldn't really make sense to call DELETE because we're not actually deleting the record, which is a fair point. I also think that it's pretty common in a lot of systems to also just mark records as deleted, so I was wondering what you think about this. Not a big deal at all, but just curious

holmesie replied:
That makes sense - HTTP DELETE doesn't _necessarily_ mean an actual database `DELETE FROM `table_name` WHERE condition;`, but I understand if the controller looks that way it could get confusing. I'm honestly okay either way, as long as it's documented/commented clearly.

timwongj replied:
thanks Adam, we can leave it the way it is then

timwongj commented on app/javascript/components/util/ConfirmDialog.js:
```diff
@@ -17,6 +17,7 @@ class Confirmation extends React.Component {
   render() {
     const {
       okLabel = 'OK',
+      okVariant = 'primary',
```
should we define `cancelVariant` as well for consistency?

sgober replied:
I don't think so - the cancel button should always look the same (as it does across the app)

timwongj commented on app/controllers/histories_controller.rb:
```diff
@@ -8,9 +8,45 @@ class HistoriesController < ApplicationController
+                          comment: params.permit(:comment)[:comment],
+                          history_type: params.permit(:type)[:type] || 'Comment')
+
+    history.original_comment = history
```
should we only populate this field for `comments` or are we thinking of having revisions for other history types in the future too?

sgober replied:
spoke with @holmesie and we want to leave this blank for non-comment histories

timwongj commented on app/controllers/histories_controller.rb:
```diff
@@ -8,9 +8,45 @@ class HistoriesController < ApplicationController
+
+    patient = current_user.viewable_patients.find_by(id: params.permit(:patient_id)[:patient_id])
+    history = patient.histories.find_by(id: params.permit(:id)[:id])
+    redirect_to root_url && return if history.nil? || history.history_type != 'Comment' || history.created_by != current_user.email
```
```suggestion
    redirect_to root_url && return if history.nil? || history.history_type != History::HISTORY_TYPES[:comment] || history.created_by != current_user.email
```

sgober replied:
fixed here: https://github.com/SaraAlert/SaraAlert/pull/793/commits/83e2b66ed498eb88bf55283e6ec1b25c43a1b2b9

timwongj commented on app/controllers/histories_controller.rb:
```diff
@@ -8,9 +8,45 @@ class HistoriesController < ApplicationController
+    redirect_to root_url && return if history.nil? || history.history_type != 'Comment' || history.created_by != current_user.email
+
+    # mark each version of the history as deleted, not just the most recent one
+    history_versions = patient.histories.where(original_comment_id: history.original_comment_id)
```
going to need to populate `original_comment_id` for existing comments during the migration

sgober replied:
added here: https://github.com/SaraAlert/SaraAlert/pull/793/commits/83e2b66ed498eb88bf55283e6ec1b25c43a1b2b9

timwongj commented on app/javascript/components/util/DeleteDialog.js:
```diff
@@ -0,0 +1,85 @@
+            <option>Other</option>
+          </Form.Control>
+          {this.state.showTextInput && (
+            <Form.Control
```
let's put a character limit on this and make sure that's validated in the back end as well

![Uploading Screen Shot 2021-05-05 at 2.54.17 AM.png…]()

timwongj replied:
@holmesie for the back end validation, would we rather add a character limit validation rule to the model or add an inline if check with `redirect_to root_url && return`

timwongj commented on app/controllers/histories_controller.rb:
```diff
@@ -8,9 +8,45 @@ class HistoriesController < ApplicationController
+  end
+
+  # "Edits" a history comment - a new history comment is created with the updated comment text and a reference to the id of the original
+  def edit
```
this is a little nit-picky, but I think it's more common to use the term `update` for an action like this (which is consistent with other controllers)

sgober replied:
I think because `update` is the default term, posting to `patient/:id/update` ends up calling `create` because it's a post.  I'd rather leave as is because of that

timwongj replied:
in that case, I updated the http method to be consistent with the action: https://github.com/SaraAlert/SaraAlert/pull/793/commits/c8f1f62444e06ae8640e77e29be166045165d6dd#diff-959bc9abc46a55332bb64d5155a79323afa75a50ec1a2137ddd22d926f62c6c5R69

timwongj commented on app/controllers/histories_controller.rb:
```diff
@@ -8,9 +8,45 @@ class HistoriesController < ApplicationController
+
+  # "Deletes" a history comment - does not actually remove the comment from the database
+  # but adds a deleted_by and deleted_reason that show the comment was deleted
+  def delete
```
this is probably not a bad time to write a `histories_controller_test.rb` file

timwongj replied:
would be nice to have more test cases, but here's a start: https://github.com/SaraAlert/SaraAlert/pull/793/commits/c8f1f62444e06ae8640e77e29be166045165d6dd#diff-e24e46e4cebd169bcec6422cfdd877e9bbfa4b402096d36398418837e351a3d5R5

sgober replied:
I fleshed these out a bit more here: https://github.com/SaraAlert/SaraAlert/pull/793/commits/37e5e22912de96896e1da829ddb1db84305d0b37. I think it's sufficient for now

timwongj commented on app/controllers/histories_controller.rb:
```diff
@@ -8,9 +8,45 @@ class HistoriesController < ApplicationController
                     created_by: current_user.email,
                     comment: params.permit(:comment)[:comment],
-                    history_type: params.permit(:type)[:type] || 'Comment')
+                    history_type: params.permit(:type)[:type] || 'Comment',
```
```suggestion
                    History::HISTORY_TYPES[:comment],
```

sgober replied:
updated here: https://github.com/SaraAlert/SaraAlert/pull/793/commits/83e2b66ed498eb88bf55283e6ec1b25c43a1b2b9

timwongj replied:
just a heads up, we still need the `|| 'Comment'` in `create` so I added it back: https://github.com/SaraAlert/SaraAlert/pull/793/commits/c8f1f62444e06ae8640e77e29be166045165d6dd#diff-c8c8197061d8db322cb11806e49c59c00413cfb234997e88adfe4e939477b69aR17

timwongj commented on app/controllers/histories_controller.rb:
```diff
@@ -8,9 +8,45 @@ class HistoriesController < ApplicationController
+    redirect_to root_url unless current_user.can_create_subject_history?
+
+    patient = current_user.viewable_patients.find_by(id: params.permit(:patient_id)[:patient_id])
+    history = patient.histories.find_by(id: params.permit(:id)[:id])
```
need to add a `nil` check for patients before trying to access `patient.histories`

sgober replied:
added here: https://github.com/SaraAlert/SaraAlert/pull/793/commits/83e2b66ed498eb88bf55283e6ec1b25c43a1b2b9

timwongj commented on app/controllers/histories_controller.rb:
```diff
@@ -8,9 +8,45 @@ class HistoriesController < ApplicationController
+    redirect_to root_url unless current_user.can_create_subject_history?
+
+    patient = current_user.viewable_patients.find_by(id: params.permit(:patient_id)[:patient_id])
+    history = patient.histories.find_by(id: params.permit(:id)[:id])
```
also need a `nil` check for `patient` here

sgober replied:
added here: https://github.com/SaraAlert/SaraAlert/pull/793/commits/83e2b66ed498eb88bf55283e6ec1b25c43a1b2b9

timwongj commented on app/controllers/histories_controller.rb:
```diff
@@ -8,9 +8,45 @@ class HistoriesController < ApplicationController
     redirect_to root_url unless current_user.can_create_subject_history?
 
-    History.create!(patient_id: params.permit(:patient_id)[:patient_id],
+    history = History.new(patient_id: params.permit(:patient_id)[:patient_id],
```
need to check if patient is viewable by user by adding
```
patient = current_user.viewable_patients.find_by(id: params.permit(:patient_id)[:patient_id])
```

timwongj replied:
https://github.com/SaraAlert/SaraAlert/pull/793/commits/c8f1f62444e06ae8640e77e29be166045165d6dd#diff-c8c8197061d8db322cb11806e49c59c00413cfb234997e88adfe4e939477b69aR11

timwongj commented on app/controllers/histories_controller.rb:
```diff
@@ -8,9 +8,45 @@ class HistoriesController < ApplicationController
     redirect_to root_url unless current_user.can_create_subject_history?
 
-    History.create!(patient_id: params.permit(:patient_id)[:patient_id],
+    history = History.new(patient_id: params.permit(:patient_id)[:patient_id],
```
some of these params (especially `patient_id` and maybe `created_by` and `comment`) should be required (also applies to the other methods as well), @holmesie which params do you think we should require?

timwongj replied:
started requiring some params such as `patient_id` and history `id` here: https://github.com/SaraAlert/SaraAlert/pull/793/commits/c8f1f62444e06ae8640e77e29be166045165d6dd#diff-c8c8197061d8db322cb11806e49c59c00413cfb234997e88adfe4e939477b69aR11

would like to know if we should require other ones as well

sgober replied:
I think this is fine since those are the only ones we use

sgober commented:
> This needs to allow users to see older versions of comments (it's okay to rely on exports for archived, but not for edited comments). I would look to GitHub comment edits for inspiration as to how it should be done.

This is now supported, just needs front end tests for the new component

sgober commented:
I think I've addressed most of the feedback here (waiting on responses from @holmesie on a few things).  The support for showing the full edit history in a modal has been added - please test!

timwongj commented on app/javascript/components/patient/history/EditHistoryModal.js:
```diff
@@ -0,0 +1,45 @@
+              .map(version => (
+                <ListGroup.Item key={version.id}>
+                  <div className="text-muted mb-1">
+                    {formatRelativePast(version.created_at)} by {version.created_by}
```
this is probably more of a question/discussion for the PO team, but should we include the absolute timestamp somehow as well?

sgober replied:
honestly, it looked way too cluttered with both.  will run it by them though

sgober replied:
only keeping the relative date, but did tweak the look on some feedback which I just pushed

DPC15038 commented on app/javascript/components/enrollment/steps/Exposure.js:
```diff
@@ -804,7 +804,7 @@ class Exposure extends React.Component {
               value={this.state.current.patient.exposure_notes || ''}
               onChange={this.handleChange}
             />
-            <Form.Label className="notes-character-limit"> {2000 - (this.state.current.patient.exposure_notes || '').length} characters remaining </Form.Label>
```
👍 Nice catch here

sgober commented on app/javascript/components/patient/history/History.js:
```diff
@@ -0,0 +1,221 @@
+    this.setState({ editMode: true });
+  };
+
+  toggleEditMode = () => {
```
definitely a mistake, updated here: https://github.com/SaraAlert/SaraAlert/pull/793/commits/d1c412f22b8cd1d15c79dbf7c80ef8b5e10523b6

DPC15038 commented on app/javascript/components/patient/history/History.js:
```diff
@@ -0,0 +1,221 @@
+import reportError from '../../util/ReportError';
+import { formatTimestamp, formatRelativePast } from '../../../utils/DateTime';
+
+const MAX_COMMENT_LENGTH = 10000;
```
I like this clean syntax 👍

sgober replied:
you came up with it ;)

DPC15038 commented on app/javascript/components/util/DeleteDialog.js:
```diff
@@ -0,0 +1,99 @@
+                value={this.state.delete_reason_text || ''}
+                onChange={this.handleTextChange}
+              />
+              <div className="character-limit-text">{MAX_REASON_LENGTH - (this.state.delete_reason_text || '').length} characters remaining</div>
```
`delete_reason_text` should be declared in the state constructor as the empty string. Cleans up this code and the line `67` as well.

DPC15038 commented on app/javascript/packs/stylesheets/layout.scss:
```diff
@@ -16,11 +16,12 @@
 }
 
-.notes-character-limit {
+.character-limit-text {
```
Good renaming here 👍

tstrass commented on db/migrate/20210430124303_add_edited_and_deleted_to_histories.rb:
```diff
@@ -0,0 +1,19 @@
+class AddEditedAndDeletedToHistories < ActiveRecord::Migration[6.1]
```
Looks like this is placed _before_ two other migrations (`RemoveContentsAndLookupFromAndAddMarkedForDeletionToDownloads` and `CreateActiveStorageTables`). I'm guessing that was just the result of a rebase, but this migration should be the most recent.

tstrass reviewed (changes requested):
A few comments but nothing that really affects the behavior. The UI for this looks great!

sgober commented on app/controllers/histories_controller.rb:
```diff
@@ -8,9 +8,53 @@ class HistoriesController < ApplicationController
+    history = patient.histories.find_by(id: params.require(:id))
+    redirect_to root_url && return if history.nil? || history.history_type != History::HISTORY_TYPES[:comment]
+
+    History.create!(patient_id: patient.id,
```
we need to use the new on create to be able to set the `original_comment_id` as `id`.  since that's not needed on edit it's easier to just use create

sgober replied:
added here: https://github.com/SaraAlert/SaraAlert/pull/793/commits/b4378bce75b0f90b57c518dc8c6d4e48c88928b6

sgober commented on app/controllers/patients_controller.rb:
```diff
@@ -24,6 +24,7 @@ def show
     @jurisdiction = @patient.jurisdiction
     @laboratories = @patient.laboratories.order(:created_at)
     @close_contacts = @patient.close_contacts.order(:created_at)
+    @histories = @patient.histories.order(:created_at).where(deleted_by: nil).group_by { |h| h.original_comment_id || h.id }.values.reverse
```
@holmesie what do you think? we added this here initially because it's grouped with the others

holmesie replied:
Why the `reverse` at the end?

tstrass replied:
>I would say this line after @patient.histories is 100% massaging the data for the front end, which makes sense to me to belong in a controller.

Fair enough. I'm fine leaving this as is

sgober commented on app/controllers/histories_controller.rb:
```diff
@@ -8,9 +8,53 @@ class HistoriesController < ApplicationController
+    redirect_to root_url && return if patient.nil?
+
+    history = patient.histories.find_by(id: params.require(:id))
+    redirect_to root_url && return if history.nil? || history.history_type != History::HISTORY_TYPES[:comment]
```
added here: https://github.com/SaraAlert/SaraAlert/pull/793/commits/db80bda73c42615becb2b4c0645aae6e92ba60f1

sgober commented on app/controllers/histories_controller.rb:
```diff
@@ -3,14 +3,54 @@
+           .update_all({ deleted_by: current_user.email, delete_reason: params.permit(:delete_reason)[:delete_reason] })
+  end
+
+  def check_role
```
added here: https://github.com/SaraAlert/SaraAlert/pull/793/commits/b4378bce75b0f90b57c518dc8c6d4e48c88928b6

sgober commented on app/controllers/histories_controller.rb:
```diff
@@ -3,14 +3,54 @@
 class HistoriesController < ApplicationController
   before_action :authenticate_user!
+  before_action :check_role
+  before_action :check_patient
```
updated here: https://github.com/SaraAlert/SaraAlert/pull/793/commits/b4378bce75b0f90b57c518dc8c6d4e48c88928b6

tstrass reviewed (approved):
LGTM

DPC15038 reviewed (approved):
Small linting fix, but besides that this looks good to me 👍 Nice job on all the hard-work

holmesie commented on app/controllers/histories_controller.rb:
```diff
@@ -2,15 +2,60 @@
+  private
+
+  def check_role
+    redirect_to root_url unless current_user.can_create_subject_history?
```
This controller is generally not handling rendering templated views, it is hit via AJAX calls (correct?). In that sense `redirect_to root_url` is probably not the best way to handle the error case. Instead, you should explicitly tell the front end this call was invalid, forbidden, etc.

Try something like:
```
respond_to do |format|
  format.any { head :forbidden }
end
```
or
```
return head :forbidden
```
This should result in the error toast popping up with the relevant error.

sgober replied:
added here: https://github.com/SaraAlert/SaraAlert/pull/793/commits/0a0fc4efb2a03ddbcdb9ba0ac01bb4adb922e3e9

holmesie reviewed (changes requested):
Overall looks great; I just have one thing I'd like to see change in the primary controller

holmesie commented on app/controllers/histories_controller.rb:
```diff
@@ -42,17 +42,17 @@ def delete
 
   def check_role
-    redirect_to root_url unless current_user.can_create_subject_history?
+    rreturn head :forbidden unless current_user.can_create_subject_history?
```
typo

sgober replied:
https://github.com/SaraAlert/SaraAlert/pull/793/commits/3df27a2434a76eed44096e2d689c2d3b01049668
